man/dbregen.1
.TH DBREGEN 1 "12 September 1995"
.FB
.SH NAME
dbregen \- regenerate indexes for a \fIFirstBase\fP database.
.SH SYNOPSIS
.B dbregen
[\fB-b\fP] [\fB-y\fP] [\fB-B\fP] [\fB-d\fP \fIdatabase\fP]
.sp 1
.B dbregen
[\fB-b\fP] [\fB-y\fP] [\fB-n\fP] [\fB-B\fP]
[\fB-d\fP \fIdatabase\fP] [ \fIindex1\fP ...]
.SH DESCRIPTION
.I dbregen
is used to globally update autoindexes and user defined indexes
for a \fIFirstBase\fP database with a single command.
The \fB-b\fP indicates batch mode and produces no
output on the terminal. The \fB-y\fP
flag runs the tool without asking
about existing files nor pausing. Each is passed to \fIdbigen\fP if needed.
.PP
If \fIdbregen\fP(1) is used the first way listed above, 
all of the autoindexes defined for that database will be updated.
Here, regeneration does not use \fIdbigen\fP(1) if possible,
but only sorts the indexes
and corrects the index header information.
.PP
If there is no actual autoindex to sort, or no index dictionary
(the \fI.idict\fP file), then \fIdbregen\fP(1)
will use \fIdbigen\fP(1)
to create the autoindex from scratch. In this case, if \fB-B\fP
is specified, it will be passed on to \fIdbigen\fP(1) for record blocking.
.PP
\fIDbregen\fP(1)
will also work on user defined indexes. This is the second form
of \fIdbregen\fP(1) listed above.
Merely provide the names
of the indexes you want \fIdbregen\fP(1) to recreate,
and \fIdbigen\fP(1) will be used to generate each index.
.PP
Note that under some circumstances, a complex index may contain all needed
entries, but not all will be sorted. This situation, normally occurring
only in autoindexes, can happen with the use of \fIdbrload\fP(1), or with
custom programs that modify a \fIFirstBase\fP index.
\fIDbregen\fP(1) will detect
this situation and merely sort the index data and update the index header.
This sort and update approach is much faster than using \fIdbigen\fP(1).
.PP
The normal regeneration of autoindexes can be suppressed by using
the \fB-n\fP option. 
.PP
\fIDbregen\fP(1) walks on the index names a little bit before processing,
taking off all extensions to the file name. This extension removal
means that shell
expansion of file names can be used to gather all \fI*.idicti\fP files,
and dbregen will correctly pass only the basenames to
the \fIdbigen\fP(1) tool.
This is quite helpful
for running shell scripts and batch jobs.
.PP
For example, the following one line command will regenerate all the
indexes in the current directory, whether an autoindex or not,
without any other keystrokes, but with visual displays:
.nf
.sp 1
   dbregen -d dbase -y *.idicti
.fi
.PP
As another example, the following commands will guarantee that all 
indexes are regenerated from scratch in batch mode.
.nf
.sp 1
   rmidx *.idx; dbregen -d dbase -b *.idicti
.fi
.SH FILES
.PD 0
dbase
default name of database.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary file.
.TP 10
*.idx
\fIFirstBase\fP index.
.PD
.SH SEE ALSO
dbigen(1), dbdind(1), dbclean(1), dbedit(1), setup(5), rmdb(8)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
Do not rebuild an autoindex that others might be using (updating).
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbregen.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbregen.c,v 9.1 2001/01/16 02:46:48 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbregen_sid[] = "@(#) $Id: dbregen.c,v 9.1 2001/01/16 02:46:48 john Exp $";
#endif

#include <fb.h>
#include <fb_vars.h>

char ddname[FB_MAXNAME];			/* for -dd named objects */
int forceflag = 0;
static char pass_env[FB_MAXLINE];
extern short int cdb_autobtree;

#if !FB_PROTOTYPES
static regen_btree();
static test_for_btree();
static int autoregen();
static int regen();
static void get_env();
/*static void usage();*/
#else /* FB_PROTOTYPES */
static int regen_btree(char *, char *, char *, char *, char *);
static int test_for_btree(char *, char *);
static int test_for_btree(char *, char *);
static int autoregen(fb_database *, char *);
static int regen(char *, char *, int);
static void get_env(char *, int, char **);
extern int main(int, char **);
/*static void usage(void);*/
#endif /* FB_PROTOTYPES */

/*
 *  dbregen - generate all the autoindexes indicated.
 *     either system to dbigen , or just sort if its an autoindex.
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         int j;
	 char flags[10], buf[FB_MAXLINE];

         (void) Dbregen_sid;
         fb_getargs(argc, argv, FB_ALLOCDB);
         get_env(pass_env, argc, argv);
         
	 ddname[0] = NULL;
	 if ((j = fb_testargs(argc, argv, "-dd")) > 0)
	    strcpy(ddname, argv[j + 1]);
         if (fb_getd_dict(cdb_db) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, cdb_db->ddict, NIL);
	 flags[0] = NULL;
	 if (!cdb_batchmode){
            fb_scrhdr(cdb_db, "ReGenerate AutoIdx");
	    fb_infoline();
	    fb_refresh();
	    }
	 else
	    strcat(flags, "-b ");
	 if (cdb_yesmode)
	    strcat(flags, "-y ");
	 if (fb_testargs(argc, argv, "-B") > 0)
	    strcat(flags, "-B ");
	 if (fb_testargs(argc, argv, "-f") > 0){
	    strcat(flags, "-f ");
            forceflag = 1;
            }
	 if (fb_testargs(argc, argv, "-n") == 0){
	    for (j = 0; j < cdb_db->nfields; j++)
	       if (cdb_db->kp[j]->aid != NULL){
		  if (regen(cdb_db->kp[j]->aid->autoname, flags, 0)==FB_ERROR)
		     if (!cdb_batchmode){
			sprintf(buf,"Cannot regenerate autoindex %s", 
			   cdb_db->kp[j]->aid->autoname);
			fb_serror(FB_MESSAGE, buf, NIL);
			}
		     }
	    }
	 for (j = 1; j < argc; j++)
	    if (equal("-d", argv[j]) || equal("-i", argv[j]))
	       j++;
	    else if (!(equal("-b", argv[j])) && 
	             !(equal("-y", argv[j])) &&
	             !(equal("-f", argv[j])) &&
		     !(equal("-n", argv[j])))
	       regen(argv[j], flags, 1);
         autoregen(cdb_db, flags);
         fb_ender();
      }

/*
 *  regen - regenerate the index for fname. if igen=1, force dbigen.
 */

   static int regen(fname, flags, igen)
      char *fname, *flags;
      int igen;
      
      {
         char buf[FB_MAXNAME], fidx[FB_MAXNAME], fidict[FB_MAXNAME];
         char fidicti[FB_MAXNAME], rname[FB_MAXNAME];
	 int fd;
	 long bsmax = 0, bsend = 0;

         if (forceflag)
            igen = 1;

	 fb_rootname(rname, fname);
	 sprintf(fidicti, "%s.idicti", fname);
	 sprintf(fidict, "%s.idict", fname);
	 sprintf(fidx, "%s.idx", fname);

         if (cdb_autobtree || test_for_btree(fname, fidicti))
            return(regen_btree(fname, rname, fidicti, fidict, flags));
	 
	 if (access(fidx, 0) == 0 && access(fidict, 0) == 0){
	    if ((fd = open(fidict, 2)) < 0)
	       return(FB_ERROR);
	    FB_CHECKERROR(fb_getxhead(fd, &bsmax, &bsend))
	    close(fd);
	    if (bsmax > bsend)		/* force a resort only */
	       igen = 0;
            else if (!igen && bsmax == bsend)
	       return(FB_ERROR);
	    }
	 /* if either idx or idict file is gone, regenerate with igen */
	 if (igen || (access(fidx, 0) != 0 || access(fidict, 0) != 0)){
	    sprintf(buf, "%s.idicti", rname);
	    if (access(buf, 0) == 0){
	       fb_scrstat(rname);
	       fb_refresh();
	       if (ddname[0] == NULL)
	          sprintf(buf, "dbigen -d %s -i %s %s",cdb_db->dbase,
                     rname, flags);
	       else
	          sprintf(buf, "dbigen -dd %s -d %s -i %s %s",
		     ddname, cdb_db->dbase, rname, flags);
               strcat(buf, pass_env); 
	       fb_system(buf, FB_WITHROOT);
	       fb_scrhdr(cdb_db, NIL); fb_infoline();
	       fb_refresh();
	       }
	    }
	 else {		/* else redo the header and sort the index (idx) */
	    if (!cdb_batchmode){
	       fb_basename(buf, fidx);
	       strcat(buf, ":Auto");
               fb_scrstat(buf);
	       fb_refresh();
	       }
	    if ((fd = open(fidict, 2)) < 0)
	       return(FB_ERROR);
	       
	    /* no checking of SEQF here - assume creation was ok. */
	    FB_CHECKERROR(fb_getxhead(fd, &bsmax, &bsend))
	    FB_CHECKERROR(fb_putxhead(fd, bsmax, bsmax))
	    close(fd);
	    sprintf(buf, "sort -o %s %s", fidx, fidx);
	    fb_system(buf, FB_NOROOT);
	    }
	 return(FB_AOK);
      }

/*
 *  regen_btree - regenerate the index for fname. if igen=1, force dbigen.
 */

   static int regen_btree(fname, rname, fidicti, fidict, flags)
      char *fname, *rname, *fidicti, *fidict, *flags;
      
      {
         char buf[FB_MAXLINE];

         if (access(fidicti, 0) == 0){
            fb_scrstat(rname);
            fb_refresh();
            if (ddname[0] == NULL)
               sprintf(buf, "dbigen -d %s -i %s %s -btree",
                  cdb_db->dbase, fname, flags);
            else
               sprintf(buf, "dbigen -dd %s -d %s -i %s %s -btree",
                  ddname, cdb_db->dbase, fname, flags);
            strcat(buf, pass_env); 
            fb_system(buf, FB_NOROOT);
            fb_scrhdr(cdb_db, NIL); fb_infoline();
            fb_refresh();
            }
	 return(FB_AOK);
      }

/*
 * test_for_btree - test the idicti and idict for btree status
 *	return either 0 or 1.
 */

   static int test_for_btree(fidicti, fidict)
      char *fidicti, *fidict;

      {
         FILE *fs;
         char line[FB_MAXLINE], lline[FB_MAXLINE];
         long bsmax, bsend;
         int ret = 0, fd;

         /* read all lines of idicti file --- if last is %\n, btree */
         fs = fopen(fidicti, "r");
         if (fs == NULL)
            return(0);

         while (fgets(line, FB_MAXLINE, fs) != NULL)
            strcpy(lline, line);
         fclose(fs);
         if (equal(lline, "%\n"))
            return(1);

         /* read all lines of idict file --- if last is %\n, btree */
         fd = open(fidict, READ);
         if (fd <= 0)
            return(0);
         fb_r_init(fd);
	 if (fb_getxhead(fd, &bsmax, &bsend) == FB_ERROR)
	    return(0);
         for(; fb_nextline(line, FB_MAXLINE) != 0; )
            if (line[0] == '%'){
               ret = 1;
               break;
               }
         close(fd);
         return(ret);
      }

   static int autoregen(db, flags)
      fb_database *db;
      char *flags;

      {
         int fd;
         char fname[FB_MAXLINE], iname[FB_MAXLINE], line[FB_MAXLINE];
         
	 fb_rootname(fname, db->ddict);
         strcat(fname, ".auto");
         if (access(fname, 0) < 0)
            return(FB_AOK);
         fd = open(fname, READ);
         if (fd <= 0){
            fb_serror(FB_CANT_OPEN, fname, NIL);
            return(FB_ERROR);
            }
         fb_r_init(fd);
         for (; fb_nextline(line, FB_MAXLINE) != 0; ){
            fb_getword(line, 1, fname);
	    fb_dirname(iname, db->ddict);
            strcat(iname, fname);
            regen(iname, flags, 0);
            }
         close(fd);
         return(FB_AOK);
      }

   static void get_env(p, argc, argv)
      char *p, *argv[];
      int argc;

      {
         int i;

         strcpy(p, " ");
         for (i = 1; i < argc; i++)
            if (strchr(argv[i], '=') != NULL){
               strcat(p, argv[i]);
               strcat(p, " ");
               }
      }

/* 
 * usage message 
 */
/*
*   static void usage()
*      {
*         fb_xerror(FB_MESSAGE,
*	    "usage: dbregen [-d dbase] [-n] [-b] [-y] [index1 ...]", NIL);
*      }
*/
